Dockerfile

```diff
@@ -59,30 +63,81 @@ RUN --mount=type=cache,target=/opt/ccache \
     CMAKE_PREFIX_PATH="$(dirname $(which conda))/../" \
     python setup.py install
 
-FROM conda as conda-installs
-ARG PYTHON_VERSION=3.11
-ARG CUDA_PATH=cu121
-ARG CUDA_CHANNEL=nvidia
+# Stage 5: Torch dependencies Installation
+FROM conda as torch-packages
 ARG INSTALL_CHANNEL=whl/nightly
-# Automatically set by buildx
-RUN /opt/conda/bin/conda update -y -n base -c defaults conda
-RUN /opt/conda/bin/conda install -y python=${PYTHON_VERSION}
-
+ARG CUDA_PATH=cu121
 ARG TARGETPLATFORM
 
-# INSTALL_CHANNEL whl - release, whl/nightly - nightly, whle/test - test channels
-RUN case ${TARGETPLATFORM} in \
-         "linux/arm64")  pip install --extra-index-url https://download.pytorch.org/whl/cpu/ torch torchvision torchaudio ;; \
-         *)              pip install --index-url https://download.pytorch.org/${INSTALL_CHANNEL}/${CUDA_PATH#.}/ torch torchvision torchaudio ;; \
-    esac && \
+# Install necessary tools for dependency extraction
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+    jq \
+    && rm -rf /var/lib/apt/lists/*
+
+# Define packages to install
+ENV PACKAGES="torch torchvision torchaudio"
+
+# Step 1: Preinstall Torch Dependencies (Cached Separately)
+RUN set -eux; \
+    # Determine PIP_PRE_FLAG based on INSTALL_CHANNEL
+    if echo "${INSTALL_CHANNEL}" | grep -q "nightly"; then \
+    PIP_PRE_FLAG="--pre"; \
+    else \
+    PIP_PRE_FLAG=""; \
+    fi; \
+    # Determine EXTRA_INDEX_URL based on TARGETPLATFORM
+    if [ "$TARGETPLATFORM" = "linux/arm64" ]; then \
+    EXTRA_INDEX_URL="https://download.pytorch.org/whl/cpu/"; \
+    else \
+    EXTRA_INDEX_URL="https://download.pytorch.org/${INSTALL_CHANNEL}/${CUDA_PATH%.}/"; \
+    fi; \
+    echo "Using EXTRA_INDEX_URL=${EXTRA_INDEX_URL} with PIP_PRE_FLAG=${PIP_PRE_FLAG}"; \
+    # Create a temporary directory to store dependencies
+    TEMP_DIR=$(mktemp -d); \
+    trap 'rm -rf "$TEMP_DIR"' EXIT; \
+    # Initialize an empty requirements file for dependencies
+    DEP_REQUIREMENTS="${TEMP_DIR}/dependencies.txt"; \
+    touch "$DEP_REQUIREMENTS"; \
+    # Iterate over each package to collect dependencies
+    for PACKAGE in $PACKAGES; do \
+    echo "Fetching dependencies for $PACKAGE"; \
+    # Fetch package metadata from PyPI
+    curl -s https://pypi.org/pypi/${PACKAGE}/json -o "${TEMP_DIR}/${PACKAGE}.json"; \
```

Comment:
This will work only when INSTALL_CHANNEL whl - release. Since data is published to PyPI

We need to add dependencies extraction for
whl/nightly - nightly
and
whl/test - test channels

One way to extract it is to read metadata file for the wheel:
download.pytorch.org/whl/nightly/cu124/torch-2.7.0.dev20250105%2Bcu124-cp310-cp310-manylinux_2_28_x86_64.whl.metadata

Reply:
Is there a common pattern from all the cases?